src/ui/Windows/MainEdit.cpp

```diff
@@ -2078,16 +2103,99 @@ void DboxMain::UpdateLastClipboardAction(const int iaction)
     m_lastclipboardaction += szTimeString;
   }
 
-  m_ilastaction = iaction;
+  m_ilastaction = cds;
   UpdateStatusBar();
 }
 
 void DboxMain::OnClearClipboard()
 {
-  UpdateLastClipboardAction(-1);
+  UpdateLastClipboardAction(ClipboardDataSource::ClearClipboard);
   ClearClipboardData();
 }
 
+void DboxMain::StartAuthCodeUpdateClipboardTimer(const pws_os::CUUID& uuidEntry)
+{
+  m_uuidEntryTwoFactorAutoCopyToClipboard = uuidEntry;
+  SetTimer(TIMER_TWO_FACTOR_AUTH_CODE_UPDATE_CLIPBOARD, std::max(USER_TIMER_MINIMUM, 1000), nullptr);
+}
+
+void DboxMain::StopAuthCodeUpdateClipboardTimer()
+{
+  m_uuidEntryTwoFactorAutoCopyToClipboard = pws_os::CUUID::NullUUID();
+  KillTimer(TIMER_TWO_FACTOR_AUTH_CODE_UPDATE_CLIPBOARD);
+}
+
+void DboxMain::OnTwoFactorAuthCodeUpdateClipboardTimer()
+{
+  if (m_uuidEntryTwoFactorAutoCopyToClipboard == pws_os::CUUID::NullUUID()) {
+    StopAuthCodeUpdateClipboardTimer();
+    return;
+  }
+
+  ItemListIter iter = Find(m_uuidEntryTwoFactorAutoCopyToClipboard);
+  ASSERT(iter != End());
+  if (iter == End()) {
+    StopAuthCodeUpdateClipboardTimer();
+    return;
+  }
+
+  CItemData& item = iter->second;
+  CItemData* pci_credential = &item;
+  if (item.IsDependent()) {
+    pci_credential = GetBaseEntry(&item);
+    ASSERT(pci_credential != NULL);
```

Comment:
What happens if the user "copies" an entry's auth code, and then deletes the entry?